system-addon/content-src/components/TopSites/TopSiteForm.jsx

```diff
@@ -84,15 +150,65 @@ export class TopSiteForm extends React.PureComponent {
     }
   }
 
+  validateCustomScreenshotUrl() {
+    const {customScreenshotUrl} = this.state;
+    return customScreenshotUrl ? this.validateUrl(customScreenshotUrl) : true;
+  }
+
   validateForm() {
-    const validate = this.validateUrl(this.state.url);
-    this.setState({validationError: !validate});
+    const validate = this.validateUrl(this.state.url) && this.validateCustomScreenshotUrl();
+
+    if (!validate) {
+      this.setState({validationError: true});
+    }
+
     return validate;
   }
 
+  _renderCustomScreenshotInput() {
+    const validationError = this.state.validationError &&
+      (this.state.screenshotRequestFailed || !this.validateCustomScreenshotUrl());
+    // Set focus on error if the url field is valid or when the input is first rendered on and is empty
+    const shouldFocus = (validationError && this.validateUrl(this.state.url)) || !this.state.customScreenshotUrl;
+
+    if (!this.state.showCustomScreenshotForm) {
+      return (<a href="" className="enable-custom-image-input" onClick={this.onEnableScreenshotUrlForm}>
+        <FormattedMessage id="topsites_form_image_enable_button" />
+      </a>);
+    }
+    return (<div className="custom-image-input-container">
+      <TopSiteFormInput
+        errorMessageId={this.props.screenshotRequestFailed ? "topsites_form_image_request_error" : "topsites_form_url_validation"}
+        loading={this.state.pendingScreenshotUpdate}
+        onChange={this.onCustomScreenshotUrlChange}
+        onClear={this.onClearScreenshotInput}
+        shouldFocus={shouldFocus}
+        typeUrl={true}
+        value={this.state.customScreenshotUrl}
+        validationError={validationError}
+        titleId="topsites_form_image_label"
+        placeholderId="topsites_form_image_placeholder"
+        intl={this.props.intl} />
+    </div>);
+  }
+
+  _getPreviewScreenshot() {
+    return Object.assign({}, this.props.site, {
+      screenshotPreview: this.state.screenshotPreview,
+      screenshotRequestFailed: this.state.screenshotRequestFailed
+    });
+  }
+
   render() {
+    const {customScreenshotUrl} = this.state;
+    const {site} = this.props;
     // For UI purposes, editing without an existing link is "add"
-    const showAsAdd = !this.props.site;
+    const showAsAdd = !site;
+    const changedCustomScreenshotUrl = site && customScreenshotUrl &&
+      (site.customScreenshotURL !== this.cleanUrl(customScreenshotUrl));
+    // Preview mode enables the preview button and prevents saving or adding a topsite.
+    const previewMode = (customScreenshotUrl && !this.state.screenshotPreview) &&
+                        (showAsAdd || changedCustomScreenshotUrl);
```

Comment:
Maybe the condition could use some refactoring. This is what I was trying to say:
* If we have a custom url and no preview it could mean that:
  * we are `edit`ing a topsite, so enable "Preview" if the value of the custom url changed
  * we are `edit`ing a topsite but no change to the url happend so show "Save"
  * we are `add`ing a topsite, we always want a "Preview" button if we don't have a preview

I couldn't find a way to include the last case in the other two (while keeping a single `previewMode` check/variable). We could further look at the site.customScreenshot value and see if it's set, but it would just replace one boolean for another.
Any thoughts?

Reply:
Maybe
```
customScreenshotUrl && !this.state.screenshotPreview && (site && !site.customScreenshot);
```
I'll looking into it some more.